pkgs/development/python-modules/matplotlib/default.nix

```diff
@@ -20,36 +19,35 @@ assert enableTk -> (tcl != null)
 assert enableQt -> pyqt5 != null;
 
 buildPythonPackage rec {
-  version = "3.1.3";
+  version = "3.2.1";
   pname = "matplotlib";
 
   disabled = !isPy3k;
 
   src = fetchPypi {
     inherit pname version;
-    sha256 = "db3121f12fb9b99f105d1413aebaeb3d943f269f3d262b45586d12765866f0c6";
+    sha256 = "ffe2f9cdcea1086fc414e82f42271ecf1976700b8edd16ca9d376189c6d93aee";
   };
 
-  NIX_CFLAGS_COMPILE = stdenv.lib.optionalString stdenv.isDarwin "-I${libcxx}/include/c++/v1";
-
   XDG_RUNTIME_DIR = "/tmp";
 
   nativeBuildInputs = [ pkgconfig ];
 
-  buildInputs = [ python which sphinx stdenv ]
+  buildInputs = [ python which sphinx ]
     ++ stdenv.lib.optional enableGhostscript ghostscript
     ++ stdenv.lib.optional stdenv.isDarwin [ Cocoa ];
 
   propagatedBuildInputs =
     [ cycler dateutil numpy pyparsing tornado freetype kiwisolver
       libpng mock pytz ]
-    ++ stdenv.lib.optional (pythonOlder "3.3") backports_functools_lru_cache
     ++ stdenv.lib.optionals enableGtk3 [ cairo pycairo gtk3 gobject-introspection pygobject3 ]
     ++ stdenv.lib.optionals enableTk [ tcl tk tkinter libX11 ]
     ++ stdenv.lib.optionals enableQt [ pyqt5 ];
 
-  patches =
-    [ ./basedirlist.patch ];
+  setup_cfg = ./setup.cfg;
```

Comment:
no need to create an environment variable, can just substitute it directly in `preBuild`

Reply:
There is a need. You can now easily `overrideAttrs` it.

Reply:
If it is intended to be overridable, then make it a parameter of the function. `overrideAttrs` should be avoided because it overrides the low-level `stdenv.mkDerivation`.

Reply:
@FRidh We certainly should not try to maintain options for every case, and, in fact, we don't even provide options for most common overrides, otherwise we would have functions defined like `{ stdenv, pname ? "matplotlib", version ? "2.3.1", src ? fetchurl { ... }, ... }:`. I thought, the accepted solution to this was that we are not standing in the way of the "low-level" plumbing and thus users can still override `src`. And I was sticking with this approach to defining my derivations and suggesting it when reviewing other PR's. This is the first time I hear that `overrideAttrs` should be avoided, because we don't really need overhead of `overrideDerivation` or `overridePythonAttrs` or a higher level functions in this particular case.

Just in case, I will try to present my original point again. If I have a derivation:
```nix
let
  foo = ./bar;
in
stdenv.mkDerivation {
  fooPhase = ''
    # ...
    cp ${foo} foo
    # ...
  '';
}
```
I am unable to do a lot of things, like easily access the value of `foo` (unless I use `<nixpkgs/path/to/bar>`, which is not overlay-friendly), or easily exchange it for something else (for example I'm not even sure if `builtins.replaceStrings` will discard the context of old value of `foo`). This is what we loose when we do it this way, but what do we gain for that?

Reply:
I see your point. We should indeed not make options for everything. The issue at least with `overrideAttrs` (and worse, `overrideDerivation`) is that overriding attributes may not always work as intended. It is this issue, not being able to tell whether it works or not without inspecting how `buildPython*` works, which is why I say, do not use `overrideAttrs`. 

Now, in this case it indeed would work fine because neither `buildPython*` nor `stdenv.mkDerivation` does anything with this specific variable. But, if this is deemed something that you want to offer that should work, I'd argue it should be part of the function interface where it is created. Does it happen often people create variables like these so they can override it later?

Reply:
> But, if this is deemed something that you want to offer that should work, I'd argue it should be part of the function interface where it is created.

I'm having trouble understanding you here. Could you, maybe, rephrase?

> Does it happen often people create variables like these so they can override it later?

People, who use overrides, will likely be able to quickly learn which kinds of expressions are easy to override and which are not. Once they have learned the good patterns, they will employ them for their contributions to nixpkgs. When I review PR's, I make suggestions based on my experience doing overrides to improve this aspect. Of course, there are many contributions out there that go against my "best practice", but that doesn't immediately invalidate my effort. And, yes, I saw there are few other people that do put $variables in their phase overrides.

Reply:
I thought I had replied but apparently not. 

> I'm having trouble understanding you here. Could you, maybe, rephrase?

The `$setup_cfg` variable is specific to `matplotlib`, not to `stdenv.mkDerivation` or `buildPythonPackage`.

Reply:
Ok, now I see what the point was.

The `setup_cfg` is not a part of an interface, but a part of implementation, just like `XDG_RUNTIME_DIR`. But we want the implementation to be conviniently overridable when we do "crazy" things (like build on a system without /tmp, or adding more stuff to setup_cfg).